WooCommerce/Classes/CardPresentPayments/CardPaymentsAdaptor.swift

```diff
@@ -0,0 +1,260 @@
+
+        paymentScreenEventPublisher = onboardingPresenterAdaptor.onboardingScreenViewModelPublisher.map { onboardingViewModel -> CardPresentPaymentEvent? in
+            guard let onboardingViewModel else {
+                return nil
+            }
+            return CardPresentPaymentEvent.showOnboarding(onboardingViewModel)
+        }
+        .merge(with: paymentScreenEventSubject)
+        .eraseToAnyPublisher()
+    }
+
+    private var paymentTask: Task<CardPresentPaymentResult, Never>?
+
+    @MainActor
+    func collectPayment(for order: Order,
+                        using discoveryMethod: CardReaderDiscoveryMethod) async -> CardPresentPaymentResult {
+        let preflightController = CardPresentPaymentPreflightController(
+            siteID: siteID,
+            configuration: CardPresentConfigurationLoader().configuration,
+            rootViewController: UIViewController(),
+            alertsPresenter: self,
+            onboardingPresenter: onboardingPresenterAdaptor,
+            bluetoothConnectionController: bluetoothConnectionController,
+            builtInConnectionController: tapToPayConnectionController)
+        let orderPaymentUseCase = CollectOrderPaymentUseCase(siteID: siteID,
+                                                                   order: order,
+                                                                   formattedAmount: currencyFormatter.formatAmount(order.total, with: order.currency) ?? "",
+                                                                   // moved from EditableOrderViewModel.collectPayment(for: Order)
+                                                                   rootViewController: UIViewController(),
+                                                                   // We don't want to use this at all, but it's currently required by the existing code.
+                                                                   // TODO: replace `rootViewController` with a protocol containing the UIVC functions we need, and implement that here.
+                                                                   onboardingPresenter: onboardingPresenterAdaptor,
+                                                                   configuration: CardPresentConfigurationLoader().configuration,
+                                                                   alertsPresenter: self,
+                                                                   preflightController: preflightController)
+        paymentTask?.cancel()
+
+        let paymentTask = Task {
+            return await withTaskCancellationHandler {
+                return await withCheckedContinuation { continuation in
+                    orderPaymentUseCase.collectPayment(using: discoveryMethod) { error in
+                        // TODO: even though we have a tri-state result type, perhaps we should throw these errors.
```

Comment:
I had a similar question when checking how we handle the collectPayment async result task within `PointOfSaleDashboardView`. Should we throw errors on `.failure`, or should we continue handling it as one of the result types of the collect payment state?

Reply:
I think trying it out and seeing what's nicer at the callsite is worthwhile on these calls... I'm not going to do it in the experiment but it would be nice to figure out our preference and be consistent.